riffdog/scanner.py

```diff
@@ -119,41 +132,99 @@ def _real_scan_element(scan_element, scanned, scanning):
     if element:
         for required_element in element.depends_on:
             if required_element not in scanned:
+                logger.info("Lookging for %s resources in real world" % required_element)
                 scanned = _real_scan_element(required_element, scanned, scanning)
 
         element.fetch_real_resources()
 
     scanned.append(scan_element)
     return scanned
 
+def _disk_inspector(state_location):
+    logger.info("File State Inspecting %s" % state_location)
+
+    target = state_location
+
+    if os.path.exists(target):
+        if os.path.isdir(target):
+            # get directories
+            for root, dirs, files in os.walk(target):
+                for filename in files:
+                    _file_inspector("%s%s" % (state_location,filename))
+
+                for direc in dirs:
+                    _disk_inspector("%s/%s/" % (state_location, direc))
+
+        if os.path.isfile(target):
+            _file_inspector(target)
+    else:
+        logger.warning("Location %s does not seem to exist" % state_location)
+    pass
+
+def _file_inspector(filename):
+    logger.info("Reading %s" % filename )
+    try:
+        with open(filename, 'r') as content_file:
+            content = content_file.read()
+        _search_state(filename, content)
+    except Exception as e:
+        # could be permission, bad file format, anything
+        logger.warn("Failed to read file - maybe not a state file for reason: %s" % e)
+
 
-def _s3_state_fetch(bucket_name):
+
+def _s3_inspector(state_location):
+    
+    logger.info("S3 State Inspecting %s" % state_location)
+    parts = state_location.split("/",1)
+    if len(parts) > 1:
+        if parts[1].endswith("/"):
+            # Fun Fact - Techincally keys can end in a / but lets assume the user isnt trying 
+            # to be awkward. - if anyone hits this, we will need to implement some kind of prefix
+            # marker so we know its a prefix or not.
+            _s3_state_fetch(parts[0], prefix=parts[1])
+        else:
+            _search_s3_state(parts[0], parts[1])
+    else:
+        _s3_state_fetch(state_location)
+
+
+def _s3_state_fetch(bucket_name, prefix=""):
     """
-    This function exists to fetch state files from S3
+    This function exists to locate state files from S3
     """
     client = boto3.client('s3')
-    s3 = boto3.resource('s3')
 
     logger.info("Processing bucket: %s" % bucket_name)
 
-    items = client.list_objects(Bucket=bucket_name, Prefix="")
+    items = client.list_objects(Bucket=bucket_name, Prefix=prefix)
 
     # _search_state(bucket_name, "sandbox_main/services/django_lambda_test/terraform.tfstate", s3)
 
     for item in items.get('Contents', []):
         if item['Key'].endswith("terraform.tfstate"): # FIXME: in future how can we better identify these files?
```

Comment:
We're still assuming `.tfstate`? Is it a separate issue to fix that?

Reply:
I've chopped it.